src/app/codegen-data-model-provider/CodegenDataModelProvider.cpp

```diff
@@ -536,6 +585,47 @@ std::optional<DataModel::ClusterInfo> CodegenDataModelProvider::GetClusterInfo(c
     return std::make_optional(std::get<DataModel::ClusterInfo>(info));
 }
 
+ClusterId CodegenDataModelProvider::FirstClientCluster(EndpointId endpointId)
+{
+    const EmberAfEndpointType * endpoint = emberAfFindEndpointType(endpointId);
+    VerifyOrReturnValue(endpoint != nullptr, kInvalidClusterId);
+    VerifyOrReturnValue(endpoint->clusterCount > 0, kInvalidClusterId);
+    VerifyOrReturnValue(endpoint->cluster != nullptr, kInvalidClusterId);
+
+    return FirstClientClusterId(endpoint, 0, mClientClusterIterationHint);
+}
+
+ClusterId CodegenDataModelProvider::NextClientCluster(const ConcreteClusterPath & before)
+{
+    const EmberAfEndpointType * endpoint = emberAfFindEndpointType(before.mEndpointId);
+    VerifyOrReturnValue(endpoint != nullptr, kInvalidClusterId);
+    VerifyOrReturnValue(endpoint->clusterCount > 0, kInvalidClusterId);
+    VerifyOrReturnValue(endpoint->cluster != nullptr, kInvalidClusterId);
+
+    if (mServerClusterIterationHint < endpoint->clusterCount)
+    {
+        const EmberAfCluster & cluster = endpoint->cluster[mServerClusterIterationHint];
+        if ((cluster.mask & CLUSTER_MASK_CLIENT) && (cluster.clusterId == before.mClusterId))
+        {
+            return FirstClientClusterId(endpoint, mClientClusterIterationHint + 1, mClientClusterIterationHint);
+        }
+    }
+
+    // linear search, this may be slow
+    // does NOT use emberAfClusterIndex to not iterate over endpoints as we have
```

Comment:
I know this got copy/pasted, but "to not iterate over endpoints" makes no sense.  What is this actually trying to say @andy31415 ?

Reply:
I have merged the two logics to one function `TryFindClusterIndex()` and add an `isServer` parameter for the function.